### DISCUSSION TITLE
cluster-operator apply KafkaConnect on the wrong cluster

### DISCUSSION THREAD

**Discussion by gmiano** [strimzi/strimzi-kafka-operator#7389] [Q&A] [answered]
**Describe the bug**
I am working on a component, made of different microservices. One of these services is a Kafka Connector on which run instances of a custom Kafka Connect that makes requests to a REST endpoint to forward messages. I use strimzi cluster operator to configure Kafka Connect through k8s resources (Kafka Connector one) instead of API requests and it watches on all namespaces.

I need to install this component multiple times on the same cluster, within different namespaces. 
kafka.strimzi.io/KafkaConnector resources are created for both installations as you can see from the screenshot below:
![image](https://user-images.githubusercontent.com/11851632/192015582-dfaeb330-90ca-4410-8ca5-ff4d5454c3fd.png)

Kafka Connect clusters have different names to avoid conflicts so I would expect cluster-operator to configure plugins only on the target cluster. However it seems that cluster operator firstly applies KafkaConnector from one namespace to both cluster, than does the same with ones from the other. It continues looping between the resources in the different namespaces and applies configs to both of them. You can look at logs for cluster-operator for insights:

[logs.txt](https://github.com/strimzi/strimzi-kafka-operator/files/9635760/logs.txt)

Below I will share the resources created on Kubernetes:

[ClusterOperator-namespace3.txt](https://github.com/strimzi/strimzi-kafka-operator/files/9636212/ClusterOperator-namespace3.txt)
[KafkaConnect-namespace1.txt](https://github.com/strimzi/strimzi-kafka-operator/files/9636213/KafkaConnect-namespace1.txt)
[KafkaConnect-namespace2.txt](https://github.com/strimzi/strimzi-kafka-operator/files/9636214/KafkaConnect-namespace2.txt)
[KafkaConnector-namespace1.txt](https://github.com/strimzi/strimzi-kafka-operator/files/9636215/KafkaConnector-namespace1.txt)
[KafkaConnector-namespace2.txt](https://github.com/strimzi/strimzi-kafka-operator/files/9636216/KafkaConnector-namespace2.txt)

It seems that cluster-operator searches for CRD resources and applies all of them to all namespaces, without discriminating on the cluster name.

**Comment by scholzj** [comment, marked as answer]
I'm not entirely sure how exactly you run things. Connectors run on top of the Connect cluster, not the other way around as you seem to suggest. 

But maybe the problem is with this statement which is not true: `Kafka Connect clusters have different names to avoid conflicts so I would expect cluster-operator to configure plugins only on the target cluster.`. This is a bit more complicated than this. The the Connect clusters connect all to the same Kafka clusters, you need to also configure the internal topics used by Connect. Otherwise they would act as one big Connect cluster. Check the docs for the details: https://strimzi.io/docs/operators/latest/full/deploying.html#con-kafka-connect-multiple-instances-str

**Comment by gmiano** [comment]
Hi @scholzj, I missed that paragraph in the official documentation. I tried it and it worked like a charm.
Thanks for your support.
